shared/chat/conversation/list-area/normal/index.desktop.js

```diff
@@ -160,74 +213,78 @@ class Thread extends React.PureComponent<Props, State> {
-      this.props.loadMoreMessages()
-    }
-  }, 100)
-
-  // When the bottom waypoint is visible, lock to bottom
-  _positionChangeBottom = debounce(({currentPosition}) => {
-    const isLockedToBottom = currentPosition === 'inside'
-    this.setState(p => (p.isLockedToBottom === isLockedToBottom ? null : {isLockedToBottom}))
-  }, 100)
-
-  _makeWaypoints = () => {
-    const waypoints = []
-    waypoints.push(
-      <TopWaypoint
-        key="topWaypoint"
-        onPositionChange={this._positionChangeTop}
-        conversationIDKey={this.props.conversationIDKey}
-      />
-    )
+  _makeItems = () => {
+    const items = []
+    items.push(<TopItem key="topItem" conversationIDKey={this.props.conversationIDKey} />)
 
     const numOrdinals = this.props.messageOrdinals.size
-    let indicies = []
-    for (var idx = 0; idx < numOrdinals; ++idx) {
-      const needNextWaypoint = idx % ordinalsInAWaypoint === 0
+    let ordinals = []
+    let previous = null
+    let lastBucket = null
+    this.props.messageOrdinals.forEach((ordinal, idx) => {
+      // We wnat to keep the mapping of ordinal to bucket fixed always
+      const bucket = Math.floor(Types.ordinalToNumber(ordinal) / ordinalsInAWaypoint)
+      if (lastBucket === null) {
+        lastBucket = bucket
+      }
+      const needNextWaypoint = bucket !== lastBucket
       const isLastItem = idx === numOrdinals - 1
       if (needNextWaypoint || isLastItem) {
         if (isLastItem) {
-          indicies.push(idx)
+          ordinals.push(ordinal)
         }
-
-        if (indicies.length) {
-          const ordinal = this.props.messageOrdinals.get(indicies[0] || 0)
-          if (!ordinal) {
-            throw new Error('Should be impossible')
-          }
-          const key = String(Types.ordinalToNumber(ordinal))
-          waypoints.push(
-            <OrdinalWaypoint key={key} id={key} rowRenderer={this._rowRenderer} indicies={indicies} />
+        if (ordinals.length) {
+          const key = String(lastBucket)
+          items.push(
+            <OrdinalWaypoint
+              key={key}
+              id={key}
+              rowRenderer={this._rowRenderer}
+              ordinals={ordinals}
+              previous={previous}
+            />
           )
-          indicies = []
+          // we pass previous so the OrdinalWaypoint can render the top item correctly
+          previous = ordinals[ordinals.length - 1]
+          ordinals = []
+          lastBucket = bucket
         }
       }
-      indicies.push(idx)
-    }
+      ordinals.push(ordinal)
+    })
 
-    waypoints.push(
-      <BottomWaypoint
-        key="bottomWaypoint"
-        onPositionChange={this._positionChangeBottom}
-        conversationIDKey={this.props.conversationIDKey}
-      />
-    )
+    items.push(<BottomItem key="bottomItem" conversationIDKey={this.props.conversationIDKey} />)
 
-    return waypoints
+    return items
+  }
+
+  _onResize = ({scroll}) => {
+    if (this._scrollHeight) {
```

Comment:
Should this do something different if `isLockedToBottom`? You can get scrolled up on making the window wider.